### ISSUE TITLE
Deprecate and migrate away from gs://kubernetes-release

### ISSUE THREAD

**Issue by spiffxp** [kubernetes/k8s.io#2396]
Part of umbrella issue to migrate the kubernetes project away from use of GCP project google-containers: https://github.com/kubernetes/k8s.io/issues/1571

This issue covers the deprecation of and migration away from the following google.com assets:
- the google.com-owned GCS bucket `gs://kubernetes-release` living in GCP project `google-containers`, in favor of the community-owned GCS bucket `gs://k8s-release` living in GCP project TBD (currently `k8s-release`)
- the region-specific GCS buckets `gs://kubernetes-release-asia` and `gs://kubernetes-release-eu`, same as above but `gs://k8s-release-eu` and `gs://k8s-release-asia` instead
- TODO: are there container images involved here as well, or did we already address that with k8s.gcr.io?

These are not labeled as steps just yet because not everything needs to be completed to full fidelity in strict sequential order.  I would prefer that we get a sense sooner rather than later what the impact of shifting dl.k8s.io traffic will be, in terms of how much budget, and what percentage of traffic that represents vs. hardcoded traffic.

### Determine new-to-deprecated sync implementation and deprecation window
There are likely a lot of people out there that have `gs://kubernetes-release` hardcoded.  It's unreasonable to stop putting new releases there without some kind of advance warning.  So after announcing our intent to deprecate `gs://kubernetes-release`, we should decide how we're going to sync new releases back there (and its region-specific buckets)
- `gsutil rsync`
- Google Cloud Storage Transfer Service
- etc.

As for the deprecation window itself, I think it's fair to treat this with a deprecation clock equivalent to disabling a v1 API.

### Determine gs://k8s-release project location and geo-sync implementation
- Someone (probably me) manually created `gs://k8s-release` and its other buckets to prevent someone else from grabbing the name
- The `-eu` and `-asia` buckets are not actually region-specific, and should be recreated as such
- We should decide how we're going to implement region syncing (same as above)
- We should decide at this stage whether we want to block on a binary artifact promotion process, or get by with one of the syncing mechanisms from above

### Use dl.k8s.io where possible and identify remaining hardcoded bucket name references across the project
The only time a kubernetes release artifact GCS bucket name needs to show up in a URI is if gsutil is involved, or someone is explicitly interested in browsing the bucket.  For tools like `curl` or `wget` that retrieve binaries via HTTP, we have `https://dl.k8s.io`, which will allow us to automatically shift traffic from one bucket to the next depending on the requested URIs

I started doing this for a few projects while working on https://github.com/kubernetes/k8s.io/issues/2318, e.g.
- https://github.com/kubernetes/cloud-provider-gcp/pull/252
- https://github.com/kubernetes-sigs/cluster-api/pull/4958

TODO: a cs.k8s.io query and resulting checklist of repos to investigate

### Shift dl.k8s.io traffic to gs://k8s-release-dev
TODO: there is a separate issue for this.  

We will pre-seed gs://k8s-release with everything in gs://kubernetes-release, and gradually modify dl.k8s.io to redirect more and more traffic to gs://k8s-release.

The idea is _not_ to  flip a switch, just in case that sends us way more traffic than our budget is prepared to handle.  Instead, let's consider shifting traffic gradually for certain URI patterns, or a certain percentage of requests, etc.  It's unclear whether this will be as straightforward as adding lines to nginx, or whether we'll want GCLB changes as well.

### Change remaining project references to gs://k8s-release

/area artifacts
/area prow
/area release-eng
/sig release
/sig testing
/wg k8s-infra
/priority important-soon
/kind cleanup
/milestone v1.23

**Comment by spiffxp** [comment]
Blocked on https://github.com/kubernetes/k8s.io/issues/1375

**Comment by ameukam** [comment]
Blocked by https://github.com/kubernetes/k8s.io/issues/4528

/milestone v1.28
/lifecycle frozen

**Comment by BenTheElder** [comment]
This isn't blocked by 4528, everything can switch to only use dl.k8s.io immediately.

**Comment by BenTheElder** [comment]
Lots of hits still https://cs.k8s.io/?q=%2Fkubernetes-release&i=nope&files=&excludeFiles=&repos=

**Comment by rjsadow** [comment]
Here are the [search results](https://cs.k8s.io/?q=https%3A%2F%2Fstorage.googleapis.com%2Fkubernetes-release&i=nope&files=&excludeFiles=vendor%2F&repos=) for what I believe are all the remaining references. I'll start creating issues with the respective repos (similar to @BenTheElder's k/website and k/test-infra issues).

**Comment by rjsadow** [comment]
Here are the results for `gs://kubernetes-release`. These changes will need to be a bit more involved and careful as:

> if those references are reading files they should switch to using https / curl / wget
> if they're writing files (kubernetes release) we can't migrate yet

- [ ] [kubernetes/kubernetes](https://github.com/kubernetes/kubernetes) | [search results](https://cs.k8s.io/?q=gs%3A%2F%2Fkubernetes-release%2F&i=nope&files=&excludeFiles=vendor%2F&repos=kubernetes/kubernetes)
- [ ] [kubernetes/cloud-provider-gcp](https://github.com/kubernetes/cloud-provider-gcp) | [search results](https://cs.k8s.io/?q=gs%3A%2F%2Fkubernetes-release%2F&i=nope&files=&excludeFiles=vendor%2F&repos=kubernetes/cloud-provider-gcp)
- [ ] [kubernetes/release](https://github.com/kubernetes/release) | [search results](https://cs.k8s.io/?q=gs%3A%2F%2Fkubernetes-release%2F&i=nope&files=&excludeFiles=vendor%2F&repos=kubernetes/release)
- [ ] [kubernetes/sig-release](https://github.com/kubernetes/sig-release) | [search results](https://cs.k8s.io/?q=gs%3A%2F%2Fkubernetes-release%2F&i=nope&files=&excludeFiles=vendor%2F&repos=kubernetes/sig-release)
- [ ] [kubernetes-sigs/kubetest2](https://github.com/kubernetes-sigs/kubetest2) | [search results](https://cs.k8s.io/?q=gs%3A%2F%2Fkubernetes-release%2F&i=nope&files=&excludeFiles=vendor%2F&repos=kubernetes-sigs/kubetest2)
- [ ] [kubernetes-sigs/tejolote](https://github.com/kubernetes-sigs/tejolote) | [search results](https://cs.k8s.io/?q=gs%3A%2F%2Fkubernetes-release%2F&i=nope&files=&excludeFiles=vendor%2F&repos=kubernetes-sigs/tejolote)
- [ ] [kubernetes/community](https://github.com/kubernetes/community) | [search results](https://cs.k8s.io/?q=gs%3A%2F%2Fkubernetes-release%2F&i=nope&files=&excludeFiles=vendor%2F&repos=kubernetes/community)
- [ ] [kubernetes/kubeadm](https://github.com/kubernetes/kubeadm) | [search results](https://cs.k8s.io/?q=gs%3A%2F%2Fkubernetes-release%2F&i=nope&files=&excludeFiles=vendor%2F&repos=kubernetes/kubeadm)

**Comment by rjsadow** [comment]
What is the plan for `gs://k8s-release-dev/ci/`? The plumbing for `https://dl.k8s.io/ci` seem to be in place already. Are folks good to start moving to that ref?

**Comment by BenTheElder** [comment]
k8s-release-dev will be tracked separately and isn't meant to be end-user facing (unlike dl.k8s.io), it's meant for contributors to the project

**Comment by BenTheElder** [comment]
> [kubernetes/release](https://github.com/kubernetes/release) | [search results](https://cs.k8s.io/?q=gs%3A%2F%2Fkubernetes-release%2F&i=nope&files=&excludeFiles=vendor%2F&repos=kubernetes/release) Cannot be updated ATT, changes are used to push artifacts to the bucket

One of them is NOT:

https://github.com/kubernetes/release/blob/065c82ea4a3ca8f0e4b1b87ade902cb9e18be78d/hack/rapture/publish-packages.sh#L60

**Comment by BenTheElder** [comment]
Even most of the release tools should be updated *unless* they're writing content, and only when writing. We'll deal with writing content later.

**Comment by BenTheElder** [comment]
Awesome, thank you! I'll plan to take a pass through remaining references again when the PRs are in.

**Comment by rjsadow** [comment]
@BenTheElder, I think we're ready for a check on outstanding references.

**Comment by ameukam** [comment]
Looking at:
- [kubernetes-client](https://github.com/search?q=%22https%3A%2F%2Fstorage.googleapis.com%2Fkubernetes-release%22+org%3Akubernetes-client&type=code)
- [kubernetes-csi](https://github.com/search?q=%22https%3A%2F%2Fstorage.googleapis.com%2Fkubernetes-release%22+org%3Akubernetes-csi&type=code)
- [kubernetes-sigs](https://github.com/search?q=%22https%3A%2F%2Fstorage.googleapis.com%2Fkubernetes-release%22+org%3Akubernetes-sigs&type=code)
- [kubernetes](https://github.com/search?q=%22https%3A%2F%2Fstorage.googleapis.com%2Fkubernetes-release%22+org%3Akubernetes&type=code)

the remaining references will go away (except the blog posts and changelogs) after the [redirect update](https://kubernetes.io/blog/2023/06/09/dl-adopt-cdn/) and the migration to a community-owned release bucket.

**Comment by BenTheElder** [comment]
Though, for the moment we're still seeing most bandwidth / requests go to the bucket.

Hopefully when we start publishing only to a new kubernetes.io bucket we'll see that start to change.

**Comment by BenTheElder** [comment]
This is ~done. After the next round of patch releases, we will drop write permissions to this old bucket.

The new bucket is private and shielded behind the CDN donated to us by Fastly, which we are using to host binary downloads going forward.

Thanks @ameukam for pushing this forward with the release engineering team and branch managers.

**Comment by rittneje** [comment]
@BenTheElder Why was this change not mentioned anywhere? We only found out when we tried to download kubectl v1.29.9 and it didn't work. This should at least have been included in the changelog.

**Comment by sftim** [comment]
We did publish https://kubernetes.io/blog/2023/08/15/pkgs-k8s-io-introduction/ @rittneje 

If you used anything other than the official source: you should have switched. We don't formally changelog infrastructure changes in the release notes, but we deliberately timed the pkgs.k8s.io article to come out on the same day as https://kubernetes.io/blog/2023/08/15/kubernetes-v1-28-release/

**Comment by sftim** [comment]
We also published https://kubernetes.io/blog/2023/06/09/dl-adopt-cdn/ earlier in 2023

**Comment by BenTheElder** [comment]
We still need to phase out the related GCB bucket and service account / project. (Which are further implemntation details of releasing). Thread in https://kubernetes.slack.com/archives/CJH2GBF7Y/p1729677695289489

**Comment by ameukam** [comment]
@BenTheElder We can drop the IAM bindings for this bucket and leave it in read-only.

**Comment by BenTheElder** [comment]
IAM bindings were dropped, this is done.

**Comment by BenTheElder** [comment]
see also https://github.com/kubernetes/k8s.io/issues/1571#issuecomment-2549990073
